File: Problems/16_3Sum_Closest.py
from typing import List

class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        nums.sort()
        closest = sum(nums[:3])
        n = len(nums)
        for i in range(n-2):
            left  = i + 1
            right = n - 1
            while left < right:
                sumUp = nums[i] + nums[left] + nums[right]
                if sumUp == target:
                    return target
                elif sumUp > target:
                    right -= 1
                else:
                    left += 1
                
                if abs(sumUp - target) < abs(closest - target):
                    closest = sumUp
        return closest


if __name__ == '__main__':
    sol = Solution()
    nums = [1,1,1,1]
    target = 4
    print(sol.threeSumClosest(nums, target))

File: Problems/16_3Sum_Closest.cpp
#include<vector>
#include<algorithm>
#include<cstdlib>

using namespace std;

class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        sort(nums.begin(), nums.end());

        int closest = 0;
        for (int i = 0; i < 3; ++i)
            closest += nums[i];
        
        for (int i = 0; i < nums.size(); ++i) {
            int l = i + 1;
            int r = nums.size()-1;
            while (l < r){
                int sumUp = nums[i] + nums[l] + nums[r];
                if (sumUp == target) {
                    return target;
                }else if (sumUp > target) {
                    --r;
                }else{
                    ++l;
                }
                if (abs(sumUp - target) < abs(closest - target))
                    closest = sumUp;
            }
        }
        return closest;
    }
};